dq.go

```
@@ -0,0 +1,212 @@
+			case int64:
+				exp = x
+			case float64:
+				exp = int64(x)
+			}
+
+			expT := time.Unix(exp, 0)
+			for _, aid := range apis.([]interface{}) {
+				apiID := aid.(string)
+
+				// This will grab the session handler
+				spec := GetSpecForApi(apiID)
+				if spec == nil {
+					log.Warning("Can't find back-end for this API, skippingc")
+					break
+				}
+
+				// Have we processed on this handler before (many APIs may use the same handler)?
+				_, processedOnSH := processedSpecs[spec.SessionManager]
+				if processedOnSH {
+					continue
+				}
+
+				// This handler hasn't been used yet for the API
+				// Get the session data
+				s, f := spec.SessionManager.GetSessionDetail(k)
+
+				// If it was found, lets process it for this handler
+				if !f {
+					// No longer in session store, delete
+					QuotaHandler.TagDelete(k)
+					continue
+				}
+
+				skip := false
+				if time.Now().After(expT) {
+					QuotaHandler.TagDelete(k)
+					skip = true
+				}
+
+				if s.IsExpired() {
+					// Remove expired data too
+					QuotaHandler.TagDelete(k)
+					skip = true
+				}
+
+				if !skip {
+					qr := int64(v.Max - v.Counter.Count())
+					if qr < 0 {
+						qr = 0
+					}
+
+					// Only write on count difference
+					if qr != s.QuotaRemaining {
+						s.QuotaRemaining = qr
+						spec.SessionManager.UpdateSession(k, s, GetLifetime(&dummyAPISpec, &s))
+						// Since we've written the token, we don't need to re-do it at the end of the middleware chain
+						s.SetFirstSeenHash()
+						// We've performed a write on this SH now, lets tag that so we don't do it again
+						processedSpecs[spec.SessionManager] = struct{}{}
+					}
+				}
+
+			}
+		})
+	}
+
+	return nil
+
+}
+
+func startDQ(statusFunc GetLeaderStatusFunc) {
+	log.WithFields(logrus.Fields{
+		"prefix": "DQuota",
+	}).Info("Using Distributed Quota")
+	p := strconv.Itoa(config.Storage.Port)
+	cs := fmt.Sprintf("redis://%v:%v", config.Storage.Host, p)
+	c1, _ := client.NewClient(cs, encoding.JSON)
```

we ignore the error here?